### ISSUE TITLE
crash when split() passed invalid argument

### ISSUE THREAD

**Issue by frobnitzem** [oilshell/oil#978]
The homebrew package "oil" crashes when running:
```
oil$ var myflags = %( --all --long )
oil$ write -- @split(myflags)
Traceback (most recent call last):
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 328, in _cpython_main_hook
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 301, in main
  File "/home/andy/git/oilshell/oil/bin/oil.py", line 292, in AppBundleMain
  File "/home/andy/git/oilshell/oil/core/shell.py", line 648, in Main
  File "/home/andy/git/oilshell/oil/core/main_loop.py", line 280, in Interactive
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1490, in ExecuteAndCatch
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 1337, in _Execute
  File "/home/andy/git/oilshell/oil/osh/cmd_eval.py", line 636, in _Dispatch
  File "/home/andy/git/oilshell/oil/osh/word_eval.py", line 1946, in EvalWordSequence2
  File "/home/andy/git/oilshell/oil/osh/word_eval.py", line 1906, in SimpleEvalWordSequence2
  File "/home/andy/git/oilshell/oil/osh/word_eval.py", line 1560, in _EvalWordToParts
  File "/home/andy/git/oilshell/oil/osh/word_eval.py", line 1542, in _EvalWordPart
  File "/home/andy/git/oilshell/oil/core/shell.py", line 402, in <lambda.1>
  File "/home/andy/git/oilshell/oil/osh/split.py", line 186, in SplitForWordEval
  File "/home/andy/git/oilshell/oil/osh/split.py", line 72, in _SpansToParts
TypeError: write() argument 1 must be string or buffer, not list
FATAL: couldn't import from app bundle '/usr/local/Cellar/oil/0.9.0/bin/oil.ovm' (1)
Stripping the oil.ovm binary may cause this error.
See https://github.com/oilshell/oil/issues/47
```

So I attempted to install from source.  However, a `#include <ctype.h>` include is missing in a python source file, as shown by:
```
$ make
build/compile.sh build-opt _build/oil/ovm-opt _build/oil/module_init.c _build/oil/main_name.c _build/oil/c-module-srcs.txt
~/dist/oil-0.9.0/Python-2.7.13 ~/dist/oil-0.9.0
Python/_warnings.c:503:13: error: implicitly declaring library function 'tolower' with type
      'int (int)' [-Werror,-Wimplicit-function-declaration]
            tolower(file_str[len-3]) == 'p' &&
            ^
Python/_warnings.c:503:13: note: include the header <ctype.h> or explicitly provide a declaration
      for 'tolower'
```

Further, several warnings are generated by incorrect use of function pointers in yajl:
```
../py-yajl/yajl.c:234:14: warning: incompatible function pointer types initializing 'PyCFunction'
      (aka 'struct _object *(*)(struct _object *, struct _object *)') with an expression of type
      'PyCFunctionWithKeywords' (aka 'struct _object *(*)(struct _object *, struct _object *,
      struct _object *)') [-Wincompatible-function-pointer-types]
    {"dump", (PyCFunctionWithKeywords)(py_dump), METH_VARARGS | METH_KEYWORDS,
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Adding the missing include, I can compile, but the make-process strips the binary, leading to the initial error again.

When I try the latest github dev version, I get a missing dependency:
```
$ make
test -d _build/opy && \
	  build/actions.sh app-deps opy ~/dist/oil:/~/dist/oil/vendor bin.opy_
build/actions.sh: line 69: _deps/cpython-full/python: No such file or directory
test -d _build/oil && \
	  build/actions.sh app-deps oil ~/dist/oil:~/dist/oil/vendor bin.oil
build/actions.sh: line 69: _deps/cpython-full/python: No such file or directory
test -d _build/hello && \
	  build/actions.sh app-deps hello ~/dist/oil/build/testdata:~/dist/oil/vendor hello
build/actions.sh: line 69: _deps/cpython-full/python: No such file or directory
```

**Comment by andychu** [comment]
Ah, FWIW this is a type mismatch issue, analogous to where Python would raise `TypeError`.

`split()` takes a string, but `%( --all --long )`, is an array, equivalent to `['--all', '--long']`.

The error is obviously bad and I'll work on a fix, although it may depend on #636

**Comment by andychu** [comment]
Released https://www.oilshell.org/release/0.9.1/
